src/VisualStudio/Core/Def/Implementation/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -0,0 +1,422 @@
+            _temporaryDirectory = Path.Combine(Path.GetTempPath(), "VisualStudioSourceGeneratedDocuments");
+            _visualStudioWorkspace = visualStudioWorkspace;
+
+            Directory.CreateDirectory(_temporaryDirectory);
+
+            _runningDocumentTable = _serviceProvider.GetService<SVsRunningDocumentTable, IVsRunningDocumentTable>();
+            _runningDocumentTableEventTracker = new RunningDocumentTableEventTracker(
+                threadingContext,
+                editorAdaptersFactoryService,
+                _runningDocumentTable,
+                this);
+        }
+
+        public void NavigateToSourceGeneratedFile(Project project, ISourceGenerator generator, string generatedSourceHintName, TextSpan sourceSpan)
+        {
+            // We will create an file name to represent this generated file; the Visual Studio shell APIs imply you can use a URI,
+            // but most URIs are blocked other than file:// and http://; they also get extra handling to attempt to download the file so
+            // those aren't really usable anyways.
+
+            // The file name we generate here is chosen to match the compiler's choice, so the debugger can recognize the files should match.
+            // We use the ProjectId as a parent directory so we can extract that out later too.
+            var generatorType = generator.GetType();
+            var temporaryFilePath = Path.Combine(_temporaryDirectory, $"{project.Id.Id}\\{generatorType.Module.ModuleVersionId}_{generatorType.FullName}_{generatedSourceHintName}");
+            Directory.CreateDirectory(Path.GetDirectoryName(temporaryFilePath));
+            File.WriteAllText(temporaryFilePath, "");
+
+            var openDocumentService = _serviceProvider.GetService<SVsUIShellOpenDocument, IVsUIShellOpenDocument>();
+            var hr = openDocumentService.OpenDocumentViaProject(
+                temporaryFilePath,
+                VSConstants.LOGVIEWID.TextView_guid,
+                out _,
+                out _,
+                out _,
+                out var windowFrame);
+
+            if (ErrorHandler.Succeeded(hr) && windowFrame != null)
+            {
+                windowFrame.Show();
+            }
+        }
+
+        public bool TryParseGeneratedFilePath(
+            string filePath,
+            [NotNullWhen(true)] out ProjectId? projectId,
+            [NotNullWhen(true)] out string? generatorTypeName,
+            [NotNullWhen(true)] out string? generatedSourceHintName)
+        {
+            if (!filePath.StartsWith(_temporaryDirectory))
+            {
+                projectId = null;
+                generatorTypeName = null;
+                generatedSourceHintName = null;
+                return false;
+            }
+
+            string fileName = Path.GetFileName(filePath);
+            string[] parts = fileName.Split(new[] { '_' }, count: 3);
+
+            generatorTypeName = parts[1];
```

Comment:
@chsienki Any idea if there's a different character we can use that's not allowed in a .NET type but is allowed in the filesystem? Maybe $ or ^ or something? The fileName.Split I'll pass a count regardless, so extra delimiters in the hint text is fine,

And I don't _have_ to do this mechanism entirely but it's somewhat handy, and I'm not sure if we're going to need this since in some cases the path may round trip to disk and back when we're restoring files between sessions.

Reply:
Looks like neither `$` or `^` are allowed in the source hint name, so so they'd work. Also good work @chsienki I couldn't break the hint name by putting `foo.xml`, it adds `.cs` to the end for me :)

Reply:
So the hint name doesn't worry me if that's last, we can just always take everything after the last 'regular' delimiter. It's knowing when the type name is done is the trick.

Reply:
We only allow any identifier character or `[.,-_ ()[]{}]` in the hint name, so anything outside of that range will work.

`$` is probably a pretty good candidate, but given that we use it as part of compiler generated names I would worry there might be a way to get the compiler to emit a type with it in the name. 

How about `+` ?

Reply:
Come to think of it, any reason we're not just naming the file with just the hint name? @chsienki we're still not writing to disk right, so the name here is just something you're making up and embedding in the PDB? That might also solve some icky issues on our side too where sometimes the debugger really wants to show the full file name in the breakpoints window which is more or less unusable. I've also gotten bitten by cases where two versions of a generator generate the same contents, but the debugger will still insist they're different because of the MVID which isn't really helping anything.

Reply:
No, unfortunately we're writing out to disk as part of another PR, where we're using the name in full. We could always construct the 'full path' hint name at output to disk time though if its easier?

Reply:
> No, unfortunately we're writing out to disk as part of another PR, where we're using the name in full.

But why not just make directories instead of a flat list of files? Is that more usable anyways?

Reply:
Hmm, we wanted a unique filename, so went that route, but actually yeah we could just use directories and retain the same differentiation. Whats the best route here, want me to make a different PR, or just contribute the change to this PR?

Reply:
@chsienki: let's do a separate PR because I'm not sure if there's other fallout yet.

Reply:
Filed #47211 to track this.